### PR TITLE
Add blocksparse support for bwd on blackwell

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

Some design decisions;
* Due to softmax pipelining in the fwd we are using a block_sparse size of 2 * tile_m and by default (at least in flex in core) we produce the bwds sparse info based off of the forward info. I added subtiling with a default of 2. So that we iterate 2 for all loaded q_indices. This can be loosened / inferred from input data and aligned w/ the triton impl so that we basically just assert sparse_q > tile_m and sparse_q%tile_m == 0 and have arbitrary subtile factor

The rest of it is basically replay of what we did in the fwd. I didnt totally extract out the compute iteration order because its a lil more mixed. And instead inlined into compute and store warps. Could refactor and think harder if needed

### Perf
combined_comparison_bwd
swa size of 2048

Perf is really good, one note is that this is comparing flash (5 dot) vs triton (7 dot) and using the 2.5 multiplier in the bwd so a little unfair but still. 

My measuring stick has been to compare causal masking implemented w/  `causal = true` vs blocksparse data. The drop from sparse iteration is much smaller than what we see in the fwd (need to take another look at the fwd)

PT changes: https://github.com/pytorch/pytorch/pull/170611

### CODE REVIEW

File: tests/cute/test_score_mod.py [hunk 1]
```diff
@@ -614,6 +614,16 @@ def score_mod_bwd_identity(grad, score, b_idx, h_idx, q_idx, kv_idx, seqlen_info
     return grad
 
 
+@cute.jit
+def score_mod_bwd_causal(grad, score, b_idx, h_idx, q_idx, kv_idx, seqlen_info, aux_tensors):
```
Comment (drisspg) [inline tests/cute/test_score_mod.py:618]: for completeness I missed in the score_mod pr

File: tests/cute/test_mask_mod.py [hunk 2]
```diff
@@ -510,6 +606,91 @@ def test_sm100_block_sparse_sink_all_masked():
...
+        q_ref = q.transpose(1, 2).requires_grad_(True)
+        k_ref = k.transpose(1, 2).requires_grad_(True)
+        v_ref = v.transpose(1, 2).requires_grad_(True)
+        grad_out_ref = grad_out.transpose(1, 2)
+
+    softmax_scale = 1.0 / math.sqrt(headdim)
+
+    # Use flex_attention directly without torch.compile for backward tests
+    # torch.compile can hang on certain mask patterns (e.g., mini_causal with float32)
+    out_ref = flex_attention(q_ref, k_ref, v_ref, block_mask=block_mask, scale=softmax_scale)
```
Comment (drisspg) [inline tests/cute/test_mask_mod.py]: I need to look into this again, this was weird but on my machine the compiled verison was basically hanging

File: tests/cute/test_mask_mod.py [hunk 3]
```diff
@@ -510,6 +606,91 @@ def test_sm100_block_sparse_sink_all_masked():
...
+        k_ref = k.transpose(1, 2).to(dtype).requires_grad_(True)
+        v_ref = v.transpose(1, 2).to(dtype).requires_grad_(True)
+        grad_out_ref = grad_out.transpose(1, 2).to(dtype)
+    else:
+        q_ref = q.transpose(1, 2).requires_grad_(True)
+        k_ref = k.transpose(1, 2).requires_grad_(True)
+        v_ref = v.transpose(1, 2).requires_grad_(True)
+        grad_out_ref = grad_out.transpose(1, 2)
+
+    softmax_scale = 1.0 / math.sqrt(headdim)
```
Comment (drisspg) [inline tests/cute/test_mask_mod.py]: remove let default handle

File: tests/cute/test_mask_mod.py [hunk 4]
```diff
@@ -510,6 +606,91 @@ def test_sm100_block_sparse_sink_all_masked():
...
+        grad_out: Gradient of output
+        mask_mod_cute: CuTE mask modification function
+        block_sparse_mask_bwd: Block sparse tensors for backward pass
+        tile_m, tile_n: Tile sizes
+        aux_tensors: Auxiliary tensors for mask_mod (e.g., doc_ids for document masking)
+
+    Returns (dq, dk, dv) all in BSHD format.
+    """
+    headdim = q.shape[-1]
+    softmax_scale = 1.0 / math.sqrt(headdim)
```
Comment (drisspg) [inline tests/cute/test_mask_mod.py]: ditto here

File: flash_attn/cute/mask.py [hunk 5]
```diff
@@ -440,17 +440,114 @@ def apply_mask_sm100_transposed(
...
         """
         assert not (mask_causal and mask_local), "mask_causal and mask_local cannot be both True"
         ROW = 0 if const_expr(not self.swap_AB) else 1
         COL = 1 if const_expr(not self.swap_AB) else 0
         assert t0ScS_t2r[0][COL] == 0, "col0 == 0"
         thr_col_offset = tScS_t2r[0][COL]
         seqlenk_col_limit = self.seqlen_k - n_block * self.tile_n - thr_col_offset
-        if const_expr(not mask_causal and not mask_local):
+
+        if const_expr(not mask_causal and not mask_local and mask_mod is not None):
```
Comment (drisspg) [inline flash_attn/cute/mask.py:469]: nit: I dont how these if els are setup really want something like
```
match mode;

blocksparse,
local/causal
_
```

**Review by tridao — APPROVED** [review]
